C# 6.0 feature: specification for null-conditional-operator

BillWagner opened the pull request:
Add the sections and grammar for the `?.` null conditional operator.

BillWagner committed:
specification for null-conditional-operator

Add the sections and grammar for the `?.` null conditional operator.

BillWagner committed:
revert an edit for another feature.

An edit from await in catch and finally clauses (#3) snuck in.

jskeet reviewed (commented):
I'm not thinking clearly enough to comment on Nigel's comments, but hopefully what I *have* commented on is still useful :)

Nigel-Ecma commented:
> Meeting March 10th: none of us can come up with a reason why this isn't a sibling of `member_access`. It would be simpler to specify that way.
> 
> Assigning to Mads to enlighten us - or we can rewrite (probably in a new PR).

Added draft PR #251 for the alternative approach

jskeet commented:
Removed meeting-discuss label as this is effectively superceded by #251 - at least, that's the hope. (If we decide not to go down that route, we'll need to do more work on this PR. If we decide we like the approach of #251, this PR can be closed.)

BillWagner commented on standard/expressions.md:
```diff
@@ -2065,7 +2066,7 @@ Within the same program, two anonymous object initializers that specify a sequen
 The `Equals` and `GetHashcode` methods on anonymous types override the methods inherited from `object`, and are defined in terms of the `Equals` and `GetHashcode` of the properties, so that two instances of the same anonymous type are equal if and only if all their properties are equal.
 
-A member declarator can be abbreviated to a simple name ([§12.7.3](expressions.md#1273-simple-names)), a member access ([§12.7.5](expressions.md#1275-member-access)) or a base access ([§12.7.9](expressions.md#1279-base-access)). This is called a ***projection initializer*** and is shorthand for a declaration of and assignment to a property with the same name. Specifically, member declarators of t ...
+A member declarator can be abbreviated to a simple name ([§12.7.3](expressions.md#1273-simple-names)), a member access ([§12.7.5](expressions.md#1275-member-access)), a base access ([§12.7.9](expressions.md#1279-base-access)), or a null-conditional member access (§null-conditional-operator-initializer). This is called a ***projection initializer*** and is shorthand for a declaration of and assign ...
```
~~It is valid, and the section [here](null-conditional-operator-initializer) describes it. I don't think an additional change is needed here.~~

Based on Jon's later comment, moving the new section here.

BillWagner commented on standard/expressions.md:
````diff
@@ -2327,6 +2328,7 @@ The `+`, `-`, `!`, `~`, `++`, `--`, cast, and `await` operators are called the u
 ```ANTLR
 unary_expression
     : primary_expression
+    | null_conditional_expression
````
Tagging @MadsTorgersen 

I think this discussion is at the heart of Nigel's alternative PR, and the discussion on `?.` being one token or two. Not making edits at this time based on this comment. I believe it accurately reflects this path toward specing the feature.

BillWagner commented on standard/expressions.md:
````diff
@@ -2340,6 +2342,106 @@ unary_expression
+
+The null-conditional operator applies a list of operations to its operand only if that operand is non-`null`. Otherwise the result of applying the operator is `null`.
+
+```antlr
````
Fixed for consistency.

BillWagner commented on standard/expressions.md:
````diff
@@ -2340,6 +2342,106 @@ unary_expression
+> ```
+> except that `a.b` and `a.b[0]` are evaluated only once. *end example*
+
+#### §null-conditional-operator-initializer Null-conditional expressions as projection initializers
````
Combined this section with 12.7.11.7 above

BillWagner replied:
The next section was moved to 13.7

BillWagner committed:
Respond to comments

BillWagner commented on standard/expressions.md:
```diff
@@ -2340,6 +2342,106 @@ unary_expression
+
+null_conditional_operations
+    : null_conditional_operations? '?' '.' identifier type_argument_list?
+    | null_conditional_operations? '?' '[' argument_list ']'
```
@Nigel-Ecma I think your concern here is addressed by the latest changes. As @jskeet pointed out two areas weren't integrated, but were written as stand alone feature changes.

BillWagner commented on standard/expressions.md:
````diff
@@ -2340,6 +2342,106 @@ unary_expression
+    ;
+```
+
+The list of operations can include member access and element access operations (which may themselves be null-conditional), as well as invocation.
````
@Nigel-Ecma I think your concern here is addressed by the latest changes. As @jskeet pointed out two areas weren't integrated, but were written as stand alone feature changes.

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -2014,13 +2014,13 @@ member_declarator_list
     ;
 
 member_declarator
```
@BillWagner - The latest change appears to replace member_declarator with only null-conditional options rather than add them to it. Has something gone awry or am I misreading this?

BillWagner replied:
@Nigel-Ecma When I walked through the definitions of the `primary_expression` and `null_conditional_operations?` that all the previous productions were included. Did I mis-interpret the ANTLR productions?

Nigel-Ecma replied:
@BillWagner – I'll take this offline and email you.

Nigel-Ecma commented on standard/statements.md:
```diff
@@ -360,18 +360,15 @@ expression_statement
     ;
 
 statement_expression
```
@BillWagner - Same as for member_declarator, here the latest change appears to replace statement_expression with only null-conditional options rather than add them to it. Has something gone awry or am I misreading this?

BillWagner committed:
fix grammar

After discussions, I went back to the text from the Microsoft C# 6 draft. The C# 6 draft put all changes in one new clause for conditional expressions. When we went to move the two grammars for anonymous object creation and member invocation, those edits introduced grammar production errors.

The only concern now is that the anonymous object creation expression does reference the null conditional expression that is defined later in the Expressions clause.

jskeet commented:
Closing as #251 was accepted instead.
